### PR TITLE
PyObject finalizer

### PR DESCRIPTION
**Pull request by amos402**
### What does this implement/fix? Explain your changes.

* Added PyObject finalizer support, Python objects referred by C# can be auto collect now.
* PythonException included C# call stack (well..., this added by debug passingly)

### Does this close any currently open issues?

...

### Any other comments?

...

### Checklist

Check all those that are applicable and complete.

-   [x] Make sure to include one or more tests for your change
-   [ ] If an enhancement PR, please create docs and at best an example
-   [ ] Add yourself to [`AUTHORS`](../blob/master/AUTHORS.md)
-   [x] Updated the [`CHANGELOG`](../blob/master/CHANGELOG.md)

### CODE REVIEW

**Comment by dmitriyse** [comment]
This is duplicate for the https://github.com/pythonnet/pythonnet/pull/532. There are several PR that are exists more than half of a year, and still not approved.

**Comment by denfromufa** [comment]
@amos402 please review #532 and let me know your feedback. @dmitriyse that code in your PR is really hard to review, it is beyond my understanding of Python/CLR internals. I still give preference to @dmitriyse PR but @amos402 if there is anything extra beyond #532 in your PR, please let us know.
I'm plan to review your PRs within 2 weeks and then merge if no issues are found.

**Comment by dmitriyse** [comment]
@denfromufa, ok. I will update those PRs in this week.

**Comment by amos402** [comment]
@denfromufa  @dmitriyse  I reviewed the #532 roughly, it create a thread for decreasing the reference of python objects, it would slow down Python if you get GIL oftenly.
But my solusion it's no need for a new python thread or .net thread. You can event use it that Python didn't compile with `WITH_THREAD` macro.
Because of Python's GIL, I suggest we should avoid use multithread on it. btw. I'm going to remove the  PyEval_InitThreads call on Initialize for my next version.

Note from python document
> Note
When only the main thread exists, no GIL operations are needed. This is a common situation (most Python programs do not use threads), and the lock operations slow the interpreter down a bit. Therefore, the lock is not created initially. This situation is equivalent to having acquired the lock: when there is only a single thread, all object accesses are safe. Therefore, when this function initializes the global interpreter lock, it also acquires it. Before the Python _thread module creates a new thread, knowing that either it has the lock or the lock hasn’t been created yet, it calls PyEval_InitThreads(). When this call returns, it is guaranteed that the lock has been created and that the calling thread has acquired it.

It is not safe to call this function when it is unknown which thread (if any) currently has the global interpreter lock.

This function is not available when thread support is disabled at compile time.

**Comment by dmitriyse** [comment]
I agree with @amos402 Py_AddPendingCall can give some benefits and my approach can be improved for Python 3.1+.
But I can agree right now with WeakReferences and GC.ReRegisterForFinalize approach. I need some research.

We are looking for different directions: single and multi threaded environment. Dedicated thread was the only approach thad succesfully solve memory leak in the my multithreaded application.

@amos402 Please check this bug on your branch, ensure that your finalizers approach also elliminates memory leaks.

**Comment by amos402** [comment]
@dmitriyse I try two cases, it seems no problems except `PythonException` was leak by original code. I fixed it on my branch.

> can be improved for Python 3.1+
Why 3.1 specificed?
`WeakReferences` only use for query garbages, it's no need for running time.
If you want to avoid call `GC.ReRegisterForFinalize`, you can just make a interface for getting all python object pointer. But I think it's not necessary, garbages will put into garbage queue, with the queue's reference, they can be resurrected.

File: src/runtime/finalizer.cs [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        }
+
+        public static readonly Finalizer Instance = new Finalizer();
+
+        public event EventHandler<CollectArgs> CollectOnce;
+
+        private ConcurrentQueue<IDisposable> _objQueue = new ConcurrentQueue<IDisposable>();
+
+        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+        private delegate int PedingCall(IntPtr arg);
```
[the remark this answers is not in the document]
Comment (amos402) [reply src/runtime/finalizer.cs]: The calling type for Py_AddPendingCall

File: src/runtime/finalizer.cs [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        public class CollectArgs : EventArgs
+        {
+            public int ObjectCount { get; set; }
+        }
+
+        public static readonly Finalizer Instance = new Finalizer();
+
+        public event EventHandler<CollectArgs> CollectOnce;
+
+        private ConcurrentQueue<IDisposable> _objQueue = new ConcurrentQueue<IDisposable>();
```
[the remark this answers is not in the document]
Comment (amos402) [reply src/runtime/finalizer.cs]: What failures will have occurred, can you give me some examples?

File: src/runtime/finalizer.cs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+        private delegate int PedingCall(IntPtr arg);
+        private readonly PedingCall _collectAction;
+
+        private bool _pending = false;
+        private readonly object _collectingLock = new object();
+        public int Threshold { get; set; }
+        public bool Enable { get; set; }
```
[the remark this answers is not in the document]
Comment (amos402) [reply src/runtime/finalizer.cs]: Make users can adjust the finalizer's configuration.

File: src/runtime/finalizer.cs [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        internal static void Shutdown()
+        {
+            if (Runtime.Py_IsInitialized() == 0)
+            {
+                Instance._objQueue = new ConcurrentQueue<IDisposable>();
+                return;
+            }
+            Instance.DisposeAll();
+            Instance.CallPendingFinalizers();
+            Runtime.PyErr_Clear();
```
[the remark this answers is not in the document]
Comment (amos402) [reply src/runtime/finalizer.cs]: Make sure no garbage left and clear the pending calls.
Pending calls wouldn't clear after calling Py_Finalize, if reset the .net environment, the function pointers would point to unknown memory and cause crash on calling them.

File: src/runtime/finalizer.cs [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+        }
+
+        private void AddPendingCollect()
+        {
+            lock (_collectingLock)
```
[the remark this answers is not in the document]
Comment (amos402) [reply src/runtime/finalizer.cs]: Why it needs to be lock whole function, the flag already set.
I can't image any situations that Py_AddPendingCall will be called in multiple threads on same time.

File: src/runtime/runtime.cs [hunk 6]
```diff
@@ -198,6 +199,8 @@ public class Runtime
         internal static bool IsPython2 = pyversionnumber < 30;
         internal static bool IsPython3 = pyversionnumber >= 30;
 
+        public static int MainManagedThreadId { get; internal set; }
```
[the remark this answers is not in the document]
Comment (amos402) [reply src/runtime/runtime.cs]: Yes, it should be. I will modify it later.

File: src/runtime/runtime.cs [hunk 7]
```diff
@@ -350,6 +354,7 @@ internal static void Initialize()
 
         internal static void Shutdown()
         {
+            Finalizer.Shutdown();
```
[the remark this answers is not in the document]
Comment (amos402) [reply src/runtime/runtime.cs]: It seems there is no particular reason for that. I can't remember why I put it on first, maybe just put it casually....

File: src/runtime/finalizer.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (filmor) [reply src/runtime/finalizer.cs]: I know, there's just a typo in the name ;)

File: src/runtime/finalizer.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (filmor) [reply src/runtime/finalizer.cs]: Why are you calling `PyErr_Clear`?

File: src/runtime/finalizer.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (filmor) [reply src/runtime/finalizer.cs]: I don't follow, either you expect multiple concurrent executions, then the lock needs to surround the whole function, or you don't, then you don't need any locking.

[the remark this answers is not in the document]
Comment (amos402) [reply src/runtime/finalizer.cs]: Oh, yes, thank you for you reminder, I neglect that there is another store at line 114.

File: src/runtime/finalizer.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (amos402) [reply src/runtime/finalizer.cs]: emmm...It seems I should expose the exceptions in DisposeAll after each `obj.Dispose()`

File: src/runtime/finalizer.cs [hunk 8]
```diff
@@ -87,24 +93,27 @@ internal static void Shutdown()
...
             Instance.CallPendingFinalizers();
-            Runtime.PyErr_Clear();
         }
 
         private void AddPendingCollect()
         {
+            if (_pending)
+            {
+                return;
+            }
```
Comment (filmor) [inline src/runtime/finalizer.cs]: Is this optimisation really needed? If not, I'd rather like the code to be obviously safe.

Comment (amos402) [reply src/runtime/finalizer.cs]: Just a simple double-check, make it check without lock at first.

**Comment by denfromufa** [comment]
@amos402 some git conflicts with master branch now.

**Comment by denfromufa** [comment]
This PR causes appveyor CI to get stuck for one hour on each build.

**Comment by amos402** [comment]
> This PR causes appveyor CI to get stuck for one hour on each build.

😥﻿I still can't figure out why it just crash on 3.7 in Linux

**Comment by amos402** [comment]
Already fixed it. It crash on 3.7 because I changed PYTHONMALLOC in test case.

**Comment by amos402** [comment]
I added ref count check on debug mode, it should be helpful for finding the bugs like this
```c#
            PyString s1 = new PyString("test_string");
            // s2 steal a reference from s1
            PyString s2 = new PyString(s1.Handle);
```

**Pull request merged**